cairo_programs/operations_with_data.cairo

```diff
@@ -0,0 +1,95 @@
+%builtins range_check bitwise 
+from starkware.cairo.common.math import signed_div_rem, unsigned_div_rem
+from starkware.cairo.common.bitwise import bitwise_operations
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.find_element import find_element, search_sorted_lower
+from starkware.cairo.common.cairo_builtins import BitwiseBuiltin
+from starkware.cairo.common.usort import usort
+
+struct GearRatios:
+    member crankset: felt
+    member cogset: felt
+end
+
+struct Bicycle:
+    member id: felt
+    member basket: felt*
+    member gear_ratios: GearRatios
+end
+
+func fill_array{range_check_ptr, bitwise_ptr: BitwiseBuiltin*}(arr: felt*, seed:felt, index: felt):
+    if index == 5:
+        return()
+    end
+    assert arr[index] = seed * index + seed
+    let (val1, val2, val3) = bitwise_operations(seed, arr[index])
+    let (q, new_seed) = unsigned_div_rem((val1 - val2 + val3 + 1000), 1000)
+    return fill_array(arr, new_seed, index + 1)
+end
+
+func initialize_bike{range_check_ptr, bitwise_ptr: BitwiseBuiltin*}(id: felt, crank: felt, cog: felt) -> (bike: Bicycle):
+    alloc_locals
+    let arr:felt* = alloc()
+    fill_array(arr, id, 0)
+    let gears = GearRatios(crankset=crank, cogset=cog)
+    return(bike=Bicycle(id=id, basket=arr, gear_ratios=gears))
+end
+
+func fill_bike_array{range_check_ptr:felt, bitwise_ptr: BitwiseBuiltin*}(arr: Bicycle*, len: felt, index: felt):
+    alloc_locals
+    if index == len:
+        return()
+    end
+
+    let (q, crank) = signed_div_rem(index + 1, 3, 30)
+    let (q, cog) = signed_div_rem(index + 2, 7, 30)
+    let (bike) = initialize_bike(index, crank, cog)
+
+    assert arr[index] = bike
+    return fill_bike_array(arr, len, index + 1)
+end
+
+func main{range_check_ptr:felt, bitwise_ptr: BitwiseBuiltin*}() -> ():
+    alloc_locals
+    let bike_arr:Bicycle* = alloc()
+    fill_bike_array(bike_arr, 30, 0)
+    let (element_ptr : Bicycle*) = find_element(
+        array_ptr=bike_arr,
+        elm_size=Bicycle.SIZE,
+        n_elms=30,
+        key=29,
+    )
+
+    assert element_ptr.gear_ratios.crankset = 0
```

Comment:
Could these assertions be made with a function that checks the assertion element by element?

Reply:
If this adds a lot of additional work don't worry

Reply:
I refactored the asserts where I could to be away from the main function. But I'm unsure if this is what you pointed out. https://github.com/lambdaclass/cleopatra_cairo/pull/362/commits/4452ab506a61988ddab8685f53d606719f2c5491